grift/src/GRIFT/Semantics/Expand.hs

```diff
@@ -1,50 +1,89 @@
+{-
+This file is part of GRIFT (Galois RISC-V ISA Formal Tools).
+
+GRIFT is free software: you can redistribute it and/or modify
+it under the terms of the GNU Affero Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+GRIFT is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero Public License for more details.
+
+You should have received a copy of the GNU Affero Public License
+along with GRIFT.  If not, see <https://www.gnu.org/licenses/>.
+-}
+
 {-# LANGUAGE DataKinds #-}
+{-# LANGUAGE ExplicitNamespaces #-}
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE GADTs #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+
+{-# OPTIONS_GHC -fplugin=GHC.TypeLits.Normalise #-}
 
 module GRIFT.Semantics.Expand where
 
 import Control.Lens ((^.))
-import Data.BitVector.Sized
-import Data.Parameterized
+import Data.Parameterized ( LeqProof(LeqProof), testLeq )
 import qualified Data.Sequence as Seq
 import Data.Sequence (Seq)
+import GHC.TypeLits ( type (<=) )
 
+import GRIFT.BitVector.BVApp
+import GRIFT.BitVector.BVFloatApp ( canonicalNaN32 )
 import GRIFT.Semantics
 import GRIFT.Semantics.Utils
-import GRIFT.Types
+import GRIFT.Types ( KnownRV, RVWidth, getSizedBV )
 
 -- | Expand an 'AbbrevApp' into its unabbreviated version.
-expandAbbrevApp :: forall expr rv w . KnownRV rv => AbbrevApp expr rv w -> expr rv w
+expandAbbrevApp ::
+  forall rv w expr.
+  BVExpr (expr rv) =>
+  -- Num (expr rv 32) =>
+  KnownRV rv =>
+  AbbrevApp expr rv w -> expr rv w
 expandAbbrevApp (SafeGPRApp wRepr ridE) = iteE
-  (ridE `eqE` litBV 0)
-  (litBV (bitVector' wRepr (0 :: Integer)))
+  (ridE `eqE` bvInteger 0)
+  (bvInteger 0)
   (stateExpr (LocApp (GPRApp wRepr ridE)))
-expandAbbrevApp (ReadCSRApp _ csr) = cases
-  [ (csr `eqE` (litBV $ encodeCSR FFlags)
-    , let fcsr = rawReadCSR (litBV $ encodeCSR FCSR)
-          flags = extractE' (knownNat @5) (knownNat @0) fcsr
-      in zextE flags
-    )
-  , (csr `eqE` (litBV $ encodeCSR FRm)
-    , let fcsr = rawReadCSR (litBV $ encodeCSR FCSR)
-          rm = extractE' (knownNat @3) (knownNat @5) fcsr
-      in zextE rm
-    )
-  ]
-  (rawReadCSR csr)
-expandAbbrevApp (NanBox32App _ e) = (litBV (-1) :: expr rv 32) `concatE` e
+expandAbbrevApp (ReadCSRApp _ csr) =
+  -- NOTE: the following code z-extends a 5-bit value, thus requiring '5 <= w'.
+  -- However, adding this as a pre-condition is a bit of a nightmare, so let's see
+  -- if we can get away with a dynamic check until we have a better solution.
+  case testLeq (knownNat @5) (knownNat @w) of
```

Comment:
This one is also awkward for the same reasons as the other `testLeq` comment on `evalInstExpr`.